File: 2-D_Dynamic_Programming/Medium/LC_62-Unique_Paths/LC_62-Unique_Paths.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // 1-D Dynamic Programming - Bottom-Up Approach
    // T: O(m*n), M: O(n), where m is num of rows, n is num of cols
    int uniquePaths(int m, int n) {
        vector<int> prevRow(n, 1);  // Solve last row base case

        for (int row = m-2; 0 <= row; row--) {
            vector<int> currRow(n, 0);
            currRow[n-1] = 1;   // Solve last col base case

            // Solve curr case using bottom up approach
            for (int col = n-2; 0 <= col; col--)
                currRow[col] = currRow[col+1] + prevRow[col];
            
            prevRow = currRow;
        }
        
        return prevRow[0];
    }

    /*
    // 2-D Dynamic Programming - Full Grid/Bottom-Up Approach
    // T&M: O(m*n), where m is num of rows, n is num of cols
    int uniquePaths(int m, int n) {
        vector<vector<int>> grid(m, vector<int>(n, 0));

        // Solve base case; last row and last cols
        for (int col = 0; col < n; col++)
            grid[m - 1][col] = 1;
        for (int row = 0; row < m; row++)
            grid[row][n-1] = 1;
        printGrid(grid);

        // Solve remaining cases using bottom up approach
        for (int row = m - 2; 0 <= row; row--) {
            for (int col = n - 2; 0 <= col; col--) {
                grid[row][col] = grid[row][col+1] + grid[row+1][col];
            }
        }
        printGrid(grid);
        
        return grid[0][0];
    }

    void printGrid(vector<vector<int>> &grid) {
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++)
                cout << grid[i][j] << " ";
            cout << endl;
        }
        cout << endl;
    }
    */
};

int main() {
    Solution sol;

    // Ex1
    assert(sol.uniquePaths(3, 7) == 28);
    // Ex2
    assert(sol.uniquePaths(3, 2) == 3);
}

File: 2-D_Dynamic_Programming/Medium/LC_62-Unique_Paths/LC_62-Unique_Paths.py
class Solution:
    # Bottom-Up Approach
    # T: O(m*n), M: O(n), where m is num of rows, n is num of cols
    def uniquePaths(self, m: int, n: int) -> int:
        prev_row = [0] * n  # Solve last row base case

        for row in range(m-1, -1, -1):
            curr_row = [0] * n
            curr_row[n - 1] = 1 # Solve last col base case

            # Solve curr case using bottom up approach
            for col in range(n-2, -1, -1):
                curr_row[col] = curr_row[col + 1] + prev_row[col]
            
            prev_row = curr_row
        
        return prev_row[0]

    '''
    # 2-D Dynamic Programming - Full Grid/Bottom-Up Approach
    # T&M: O(m*n), where m is num of rows, n is num of cols
    def uniquePaths(self, m: int, n: int) -> int:
        grid = [[0] * n for ln in range(m)]

        # Solve base case for last row & last cols
        for row in range(m):
            grid[row][n-1] = 1
        for col in range(n):
            grid[m-1][col] = 1
        print(grid)

        # Solve for the rest of the grid using bottom-up approach
        for row in range(m-2, -1, -1):
            for col in range(n-2, -1, -1):
                grid[row][col] = grid[row+1][col] + grid[row][col+1]
        print(grid)

        return grid[0][0]
    '''

sol = Solution()

# Ex1:
attempt = sol.uniquePaths(m = 3, n = 7)
assert attempt == 28, f'Expected {28}, but got {attempt}'
# Ex2:
attempt = sol.uniquePaths(m = 3, n = 2)
assert attempt == 3, f'Expected {3}, but got {attempt}'
